### PR TITLE
doc: major update to acess.md

### PR DESCRIPTION
**Pull request by rvagg**
Coming out of #2005 and a discussion amongst infra via email, I thought it would be good to put down more of the reasoning behind the levels of access so it's clear to us what that is and hopefully clear to all how we deal with granting additional access (and that there is reasoning and a path for individuals who invest their time and skills here).

Some big changes but my hope is that this is you can be told to read if you want to become a member and it will all be clear how we work organisationally.

The bottom line is that we need to expand our people resources here and some of that expansion involves elevated access. But the path to elevated access is _really_ tricky because we're dealing with security, stability, legal concerns and even concerns that are difficult to quantify like keeping our infra donors happy and not stepping on toes.

### CODE REVIEW

File: doc/access.md [hunk 1]
```diff
@@ -1,49 +1,162 @@
...
+  this repository where you see need or attempting to assist Build WG members
+  where you are able. Competence may also be demonstrated through contributions
+  to other activities of the Node.js project, although competence in software
+  development does not necessarily correlate with the type of technical
+  competence the Build WG would prefer.
+* Trust: because we are granting access to protected resources, the Build WG
+  needs to establish trust. Individuals with no prior relationship to the
+  Node.js project or one of its member companies are likely to be asked to
+  contribute as a non-member where possible for a period of time to establish
+  the basics of a trust relationship. The most two most straigtforward paths to
```
Comment (joaocgreis) [inline doc/access.md]:
```suggestion
  the basics of a trust relationship. The most two most straightforward paths to
```

File: doc/access.md [hunk 2]
```diff
@@ -1,49 +1,162 @@
...
+existing members to let us know. The best means for communication with the Build
+WG are either here via GitHub issues or through the `#node-build` channel on
+Freenode [IRC][].
+
+Membership is granted by way of a pull request adding a new individual to the
+members list on the README (e.g. [#524][]). New members can open such a pull
+request themselves, but it would be advisable to check with an existing member
+before doing so. Alternatively, an existing member may "sponsor" a new member by
+opening a pull request. Existing, active members will vote on a pull request and
+where no objetions are present after a reasonable period of time, membership
```
Comment (joaocgreis) [inline doc/access.md]:
```suggestion
where no objections are present after a reasonable period of time, membership
```

File: doc/access.md [hunk 3]
```diff
@@ -1,49 +1,162 @@
...
+will be granted.
+
+Depending on the level of trust and competence assessed by existing Build WG
+members, new members may not be given immediate access to protected resources.
+This is at the discretion of the Build WG. We ask that you not take offence if
+we appear slow to grant access to resources you ask for as we take our
+responsibilities regarding security and the stability of our infrastrucutre very
```
Comment (joaocgreis) [inline doc/access.md]:
```suggestion
responsibilities regarding security and the stability of our infrastructure very
```

File: doc/access.md [hunk 4]
```diff
@@ -1,49 +1,162 @@
...
+one or more technical areas covered by the Build WG such that the addition of
+a new member spreads the maintenance burden rather than creating additional
+burden because of lack of competence, or un-trustworthy behavior. The Build WG
+is not an expert group and we offer a place to learn and develop skills. Members
+should be aware of the bounds of their expertise and act accordingly.
+
+* Competence: it is difficult to objectively gauge technical competence without
+  demonstration. You can demonstrate competence by contributing to resources in
+  this repository where you see need or attempting to assist Build WG members
+  where you are able. Competence may also be demonstrated through contributions
```
Comment (sam-github) [inline doc/access.md]: I'm very, very sympathetic to the risks of incautious access, but I worry that this veers close to a catch-22. Despite that sense that we have an intimidating mountain of work available, I'm struggling to find helpful things that can be done without access, and without doing helpful things, its hard to demonstrate competence. @richardlau and I both demonstrated competence the long way: nodejs collaboration.

Comment (richardlau) [reply doc/access.md]: There are a few things that can be contributed without access, for example I have [four pull requests](https://github.com/nodejs/build/pulls?q=is%3Apr+author%3Arichardlau+is%3Aclosed) merged into this repository that happened before I was given access ("contributing to resources in this repository where you see need"). Since collaborators can view (but not change) job configuration in Jenkins it's even possible to [suggest fixes](https://github.com/nodejs/node/issues/24349#issuecomment-440490499) to jobs ("attempting to assist Build WG members where you are able").

I wouldn't necessarily say it's easy -- you need some idea of what is where.

Comment (rvagg) [reply doc/access.md]: Agree, this is a very difficult corner but this PR is intended to clarify current thinking and open discussion for amending that thinking where we can. So if you have suggestions for how to change things then speak up!

**Comment by sam-github** [comment]
This PR reflects how it is, so I'm good with landing it.

I'm not sure if here or https://github.com/nodejs/build/issues/1990 or somewhere else is a better place to discuss this, but I'll do it here for now.  For bystanders, I don't have a handy link other than the last WG meeting, but the TSC is asking the board to ask the member companies to contribute people to the build WG.


I am concerned that it will be difficult for anyone who has time donated to the project to get enough access to actually do anything with the current policy.

At the same time, I don't want to suggest a policy of "instant access for any IBM employees", that would be unreasonable. I hope that is obvious, but I want to acknowlege that that would be totally inappropriate, and I am not trying to suggest such a thing.

But, for example, if any active WG member (@joaocgreis or @rvagg as an example) introduced someone I'd never heard of before, provided evidence they fit the "trust through employer criteria", and personally committed to mentor the individual into the WG, review their work, help them understand how the systems work, etc., I would give them a plus 1 for basic access `*`.

I'm willing to be told that I'm too generous with my transitive trust, and should be more cautious, but that's how I feel at the moment.

`*`: start/stop ci jobs and view ci config, access to view secrets repo and access to decrypt the "test" secrets -- so basically ssh access into test machines, which is sufficient to ansible them, start jobs to test correctness of machine config, and to land things in the build WG repo.

**Comment by sam-github** [comment]
Somewhere else it was suggested I could say something about how IBM manages sensitive access, in case it gives any useful context. I don't speak with authority on our policy, but what it looks like to me as a user is:

1. If I want access to a resource, there will be a website I go to to make the request. I'll be given a link to the policy for that resource, and asked why I want it, and I submit the request and reason for it.
2. (I think) behind the scenes my manager certifies my request reason is accurate, and the manager of the resource certifies (or not!) that my request meets the policy.
3. I get access, often an email with a OTP and link to instructions.
4. When I don't want it anymore, I can go back the site, and request removal. I think this usually needs less or no approval.
5. ***important*** it seems most or all resources trigger a yearly reminder to re-certify that the resource is still in use and needed, at which point if I don't reconfirm the access request, my access gets automatically removed

So the basic principle seems to be chain of trust between humans, backed up by record keeping so its always known who has access and who authorized it, and automated systems to add and remove access, and particularly to remove access after a year if its not re-requested.
